### DISCUSSION TITLE
Regular brew maintenance

### DISCUSSION THREAD

**Discussion by jasonkarns** [Homebrew/discussions#4542] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.20
ORIGIN: https://github.com/Homebrew/brew
HEAD: 8917acac9560f8e2846eedc4181470f0345f6db8
Last commit: 6 hours ago
Core tap JSON: 30 May 13:19 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CACHE: /Users/jasonkarns/.cache/Homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_NO_ANALYTICS: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.40.1 => /opt/homebrew/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  rake-completion
```

### Description of issue

This is intended as a question/discussion, not an issue.

I have had a script (alias, technically) that I run regularly (weekly or semi-monthly) for quite a while. Its purpose is to keep my system dependencies up to date (and so that if I encounter issues, the disparity of versions from the upgrade are smaller and easier to debug).

With all the (somewhat) recent changes to brew's behavior, I'm curious what the brew community's braint trust thinks of the current command. Anything that's redundant or could be simplified?
Anything missing the folks would recommend or do as part of your own regular cleanup/maintenance?

```sh
brew cleanup ; brew doctor || confirm && brew outdated && brew upgrade
```

Brew doctor runs which _may_ fail, so I have a 'confirm' step that blocks until I either kill the command or allow it to proceed. This gives me a chance to read the doctor output and make a decision.
Then I print the outdated formulae (just for visibility) and upgrade everything.

One known limitation of the current script is that pinned formulae cause the thing to fail. I'm pretty sure it's `brew-upgrade` that is failing when anything is pinned (which is a bit frustrating; it seems wrong to me that using brew's pin feature means `brew upgrade` will fail forever, but okay)

Any thoughts or feedback?

**Comment by MikeMcQuaid** [comment]
I use `brew bundle --cleanup --global --force` for this.

**Reply by jasonkarns** [reply]
Oh I like this idea.

**Comment by carlocab** [comment]
You could probably do something like
```zsh
comm -23 <(brew outdated | sort) <(brew list --pinned | sort) | xargs brew upgrade
```
to try to upgrade only the outdated formulae that are not pinned. Piping to `sort` may also not be needed; I forget whether `brew` already sorts the output by default.

**Reply by jasonkarns** [reply]
After some digging, I learned `--ignore-pruned` is a thing, so I've just added that to the upgrade command. And with brew's automatic cleanup these days, I dropped the explicit brew cleanup step.

I did add `brew autoremove` to the end now... Would be neat if that were an option as part of upgrade ;)

**Reply by carlocab** [reply]
> Would be neat if that were an option as part of upgrade ;)

Not quite part of `brew upgrade`, but you may be interested in `HOMEBREW_AUTOREMOVE`.
